pkg/query-service/app/parser_test.go

```diff
@@ -862,3 +860,353 @@ func TestParseQueryRangeParamsPromQLVars(t *testing.T) {
+						Expression:         "B",
+					},
+					"F1": {
+						QueryName:  "F1",
+						Expression: "A/B",
+					},
+				},
+			},
+			expectErr: true,
+			errMsg:    "Group keys must match or be a subset of the other but found left: [service_name operation_name], right: [operation_name status_code]",
+		},
+		{
+			desc: "Nested Expressions with Matching Keys - Testing expressions that involve operations (e.g., addition, division) with series whose keys match or are subsets.",
+			compositeQuery: v3.CompositeQuery{
+				PanelType: v3.PanelTypeGraph,
+				QueryType: v3.QueryTypeBuilder,
+				BuilderQueries: map[string]*v3.BuilderQuery{
+					"A": {
+						QueryName:          "A",
+						DataSource:         v3.DataSourceMetrics,
+						AggregateOperator:  v3.AggregateOperatorSum,
+						AggregateAttribute: v3.AttributeKey{Key: "signoz_calls_total"},
+						GroupBy:            []v3.AttributeKey{{Key: "service_name"}, {Key: "operation_name"}},
+						Expression:         "A",
+					},
+					"B": {
+						QueryName:          "B",
+						DataSource:         v3.DataSourceMetrics,
+						AggregateOperator:  v3.AggregateOperatorSum,
+						AggregateAttribute: v3.AttributeKey{Key: "signoz_calls_total"},
+						GroupBy:            []v3.AttributeKey{{Key: "service_name"}, {Key: "operation_name"}},
+						Expression:         "B",
+					},
+					"F1": {
+						QueryName:  "F1",
+						Expression: "A + B",
+					},
+				},
+			},
+			expectErr: false,
+		},
+		{
+			desc: "Nested Expressions with Matching Keys - Testing expressions that involve operations (e.g., addition, division) with series whose keys match or are subsets.",
+			compositeQuery: v3.CompositeQuery{
+				PanelType: v3.PanelTypeGraph,
+				QueryType: v3.QueryTypeBuilder,
+				BuilderQueries: map[string]*v3.BuilderQuery{
+					"A": {
+						QueryName:          "A",
+						DataSource:         v3.DataSourceMetrics,
+						AggregateOperator:  v3.AggregateOperatorSum,
+						AggregateAttribute: v3.AttributeKey{Key: "signoz_calls_total"},
+						GroupBy:            []v3.AttributeKey{{Key: "service_name"}, {Key: "operation_name"}},
+						Expression:         "A",
+					},
+					"B": {
+						QueryName:          "B",
+						DataSource:         v3.DataSourceMetrics,
+						AggregateOperator:  v3.AggregateOperatorSum,
+						AggregateAttribute: v3.AttributeKey{Key: "signoz_calls_total"},
+						GroupBy:            []v3.AttributeKey{{Key: "service_name"}},
+						Expression:         "B",
+					},
+					"C": {
+						QueryName:          "C",
+						DataSource:         v3.DataSourceMetrics,
+						AggregateOperator:  v3.AggregateOperatorSum,
+						AggregateAttribute: v3.AttributeKey{Key: "signoz_calls_total"},
+						GroupBy:            []v3.AttributeKey{{Key: "service_name"}, {Key: "operation_name"}, {Key: "status_code"}},
+						Expression:         "C",
+					},
+					"F1": {
+						QueryName:  "F1",
+						Expression: "C/(A + B)",
```

Comment:
A+B doesn't make sense in this example?

Reply:
I feel A+B always should match groupBy keys. Any example where they are different?

Reply:
I agree the arithmetic doesn't make sense. We are not interested in that. This is testing whether it is possible to join them. The `expectErr: true` cases include the ones where they are different.

Reply:
I see that but not able to make sense why this test case has `expectErr: false` but not for https://github.com/SigNoz/signoz/pull/4402/commits/535eaa07f35fc15eb60066fefad2c9c9f81fb48e#diff-c92468e920f66a6614fa0b6c5b37f368faec8c9514790ae182f2807adcbc8bd1R1110

Reply:
The logic is simple. Either fully match or subset. `[operation_name status_code]` is not a subset of `[service_name operation_name]`

Reply:
If these were tables, You wouldn't be able to join the tables because the `status_code` is missing from the left table `A` and `service_name` is missing from the right table `B`

Reply:
It's easier to see it this way